components/codec/src/byte.rs

```diff
@@ -344,50 +352,55 @@ impl MemComparableByteCodec {
         let src_ptr_untouched = src_ptr;
         let dest_ptr_untouched = dest_ptr;
 
-        unsafe {
-            let src_ptr_end = src_ptr.add(src_len);
+        let src_ptr_end = src_ptr.add(src_len);
 
-            loop {
-                let src_ptr_next = src_ptr.add(MEMCMP_GROUP_SIZE + 1);
-                if std::intrinsics::unlikely(src_ptr_next > src_ptr_end) {
-                    return Err(ErrorInner::eof().into());
-                }
+        loop {
+            // Make sure the source buffer is the expected length before
+            // constructing the next source pointer. Note that it is UB to
+            // even _create_ a pointer that points beyond its allocation.
+            let ptrdiff = src_ptr_end.offset_from(src_ptr);
+            let ptroob = ptrdiff < MEMCMP_GROUP_SIZE as isize + 1;
 
-                // Copy `MEMCMP_GROUP_SIZE` bytes any way. However we will truncate the returned
-                // length according to padding size if it is the last block.
-                std::ptr::copy(src_ptr, dest_ptr, MEMCMP_GROUP_SIZE);
+            if std::intrinsics::unlikely(ptroob) {
+                return Err(ErrorInner::eof().into());
+            }
 
-                let padding_size = T::parse_padding_size(*src_ptr.add(MEMCMP_GROUP_SIZE));
-                src_ptr = src_ptr_next;
-                dest_ptr = dest_ptr.add(MEMCMP_GROUP_SIZE);
+            let src_ptr_next = src_ptr.add(MEMCMP_GROUP_SIZE + 1);
 
-                // If there is a padding, check whether or not it is correct.
-                if std::intrinsics::unlikely(padding_size > 0) {
-                    // First check padding size.
-                    if std::intrinsics::unlikely(padding_size > MEMCMP_GROUP_SIZE) {
-                        return Err(ErrorInner::bad_padding().into());
-                    }
+            // Copy `MEMCMP_GROUP_SIZE` bytes any way. However we will truncate the returned
```

Comment:
Is this comment still true? It sounds like it is invalidated by the changes.

Reply:
I believe the comment is true and is unrelated to the error-handling path here. By my reading, padding can be applied to any record entry, and this is just saying to "copy all the padding, we'll truncate it out later".